### PR TITLE
[PVR] EPG persistence speed improvements and fixes

### LINKED ISSUE #17800: Long shutdown time and intensive HDD usage updating EPG for large IPTV playlists
### Describe the bug

Kodi creates and deletes the `Epg*.db-journal` for every EPG item being persisted. This causes excessive HDD wear and long shutdown times (2-5 minutes) if IPTV playlist and EPG have several thousands of live TV channels and VODs.

### Expected Behavior

The journal files for `Epg*.db` should be processed in memory so shutdown times are instant (~5-8 seconds on the same playlist / EPG pair vs the unmodified setup)

### Actual Behavior

The iotop graph is shown below:

iotop-excessive-hdd-usage

The shutdown time is very long (2-5 minutes on AMD E-300 dual-core CPU and 5400rpm HDD)

### Possible Fix

I created the script and the systemd unit mounting the `$KODIPROFILE/userdata/Database` as an overlayfs with the lower directory of original `Database` directory and `tmpfs` mount as upper-directory:

[kodi-mount-overlay.sh](https://github.com/xbmc/xbmc/files/4573561/kodi-mount-overlay.sh.txt)
[kodi-mount-overlay.service](https://github.com/xbmc/xbmc/files/4573566/kodi-mount-overlay.service.txt)

With this ugly hack, the shutdown time and disk usage is desirable.

### To Reproduce

1. Generate the test playlist / EPG pair using the script:
[generate-big-test-m3u-epg.sh](https://github.com/xbmc/xbmc/files/4573573/generate-big-test-m3u-epg.sh.txt)

2. Load the M3U and EPG into IPTV Simple Client and let Kodi process the channels and EPG markers
3. Exit Kodi and observe high disk usage running `sudio iotop` and lots of created-and-vanishing `Epg36.db-journal` in `$KODIPROFILE/userdata/Database`

### Debuglog

The debuglog can be found here:

[kodi.log](https://github.com/xbmc/xbmc/files/4573629/kodi.log)

The log ended after I had to issue
``` shell
kill -9 $(pidof kodi-x11)
```

on my test VM.

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [x] Linux
 - [ ] OSX
 - [ ] Raspberry-Pi
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: Debian 10 buster amd64
 - Kodi version: 19.0 trunk

### PR DESCRIPTION
**Pull request by ksooo**
Should eventually fix #17800.

Runtime-tested on my test machine, by @emveepee and @mglae - all confirmed sufficient speed now.

@phunkyfish mind taking a look at the code changes?

### CODE REVIEW

File: xbmc/dbwrappers/dataset.h [hunk 1]
```diff
@@ -469,8 +471,13 @@ class Dataset  {
 /*clear delete_sql*/
   void clear_delete_sql();
 
-/*get value of select_sql*/
- const char *get_select_sql();
+  /* size of insert_sql*/
+  size_t insert_sql_count();
```
Comment (phunkyfish) [inline xbmc/dbwrappers/dataset.h:475]: Is there a reason to use this formatting for the function name? Why not `InsertSqlCount()`?

File: xbmc/dbwrappers/dataset.h [hunk 2]
```diff
@@ -469,8 +471,13 @@ class Dataset  {
...
+  /* size of delete_sql*/
+  size_t delete_sql_count();
```
Comment (phunkyfish) [inline xbmc/dbwrappers/dataset.h:477]: Is there a reason to use this formatting for the function name? Why not `DeleteSqlCount()`?

File: xbmc/dbwrappers/dataset.h [hunk 3]
```diff
@@ -469,8 +471,13 @@ class Dataset  {
...
+
+  /*get value of select_sql*/
+  const char* get_select_sql();
```
Comment (phunkyfish) [inline xbmc/dbwrappers/dataset.h:480]: Is there a reason to use this formatting for the function name? Why not `GetSelectSql()`?

**Review by phunkyfish — COMMENTED** [review]
Just some queries on function naming.

File: xbmc/dbwrappers/dataset.h (hunk 1 above)
Comment (ksooo) [reply xbmc/dbwrappers/dataset.h:475]: I followed the style used in the rest of the header file.

File: xbmc/dbwrappers/dataset.h (hunk 3 above)
Comment (ksooo) [reply xbmc/dbwrappers/dataset.h:480]: See my previous reply.

File: xbmc/dbwrappers/dataset.h (hunk 2 above)
Comment (ksooo) [reply xbmc/dbwrappers/dataset.h:477]: See my previous reply.

File: xbmc/dbwrappers/dataset.h (hunk 1 above)
Comment (phunkyfish) [reply xbmc/dbwrappers/dataset.h:475]: Fair

**Review by phunkyfish — APPROVED** [review]

File: xbmc/pvr/epg/EpgContainer.cpp [hunk 4]
```diff
@@ -312,14 +312,26 @@ bool CPVREpgContainer::PersistAll(unsigned int iMaxTimeslice) const
...
-        bReturn &= epg->Persist(database, true);
+        bReturn &= epg->QueuePersistQuery(database);
+
+        size_t queryCount = database->GetInsertQueriesCount() + database->GetDeleteQueriesCount();
+        if (queryCount > 10000)
+        {
+          CLog::Log(LOGDEBUG, LOGEPG, "EPG Container: committing %d queries in loop.", queryCount);
+          database->CommitDeleteQueries();
+          database->CommitInsertQueries();
+          CLog::Log(LOGDEBUG, LOGEPG, "EPG Container: committed %d queries in loop.", queryCount);
```
Comment (AlwinEsch) [inline xbmc/pvr/epg/EpgContainer.cpp:323]: About new logs can be the `{}` used. And maybe with different strings, on first view I thought both are the same :sweat_smile:. Also not sure if the `queryCount` not changed, to write again.

Maybe?:
```cpp
          CLog::Log(LOGDEBUG, LOGEPG, "EPG Container: Start committing {} queries in loop", queryCount);
          database->CommitDeleteQueries();
          database->CommitInsertQueries();
          CLog::Log(LOGDEBUG, LOGEPG, "EPG Container: End committed queries in loop");
```

Comment (ksooo) [reply xbmc/pvr/epg/EpgContainer.cpp:323]: This id for another day/PR - when I find the time to change whole PVR component to use the new logging syntax.

**Pull request merged**